Implement pack/unpack helpers

jessegrabowski opened the pull request:
<!-- !! Thank your for opening a PR !! -->

<!--- Provide a self-contained summary of your changes in the Title above -->
<!--- This is what will be shown in the automatic release notes: https://github.com/pymc-devs/pytensor/releases -->

## Description
<!--- Describe your changes in detail -->
Adds pt.pack and pt.unpack helpers, roughly conforming to the `einops` functions of the same name.

These helps are for situations where we have a ragged list of inputs that need to be raveled into a single flat list for some intermediate step. This occurs in places like optimization.

Example usage:

```py
x = pt.tensor("x", shape=shapes[0])
y = pt.tensor("y", shape=shapes[1])
z = pt.tensor("z", shape=shapes[2])

flat_params, packed_shapes = pt.pack(x, y, z)
```

Unpack simply undoes the computation, although there's norewrite to ensure `pt.unpack(*pt.pack(*inputs))` is the identity function:

```py
x, y, z = pt.unpack(flat_params, packed_shapes)
```

The use-case I forsee is creating replacement for a function of the inputs we're packing, for example:

```py
loss = (x + y.sum() + z.sum()) ** 2

flat_packed, packed_shapes = pack(x, y, z)
new_input = flat_packed.type()
new_outputs = unpack(new_input, packed_shapes)

loss = pytensor.graph.graph_replace(loss, dict(zip([x, y, z], new_outputs)))
fn = pytensor.function([new_input], loss)
```

Note that the final compiled function depends only on `new_input`, only because the shapes of the 3 packed variables were statically known. This leads to my design choices section:

1. I decided to work with the static shapes directly if they are available. This means that `pack` will eagerly return a list of integer shapes as `packed_shapes` if possible. If not possible, they will be symbolic shapes. This is maybe an anti-pattern -- we might prefer a rewrite to handle this later, but it seemed easy enough to do eagerly.
2. I didn't add support for batch dims. This is left to the user to do himself using `pt.vectorize`. 
3. The `einops` API has arguments to support packing/unpacking on arbitrary subsets of dimensions. I didn't do this, because I couldn't think of a use-case that a user couldn't get himself using `DimShuffle` and `vectorize`.

## Related Issue
<!--- It is good practice to first open an issue explaining the bug / new feature that is addressed by this PR -->
<!--- Please type an `x` in one of the boxes below and provide the issue number after the # sign: -->
- [ ] Closes #1553 
- [ ] Related to #1550 

## Checklist
<!--- Make sure you have completed the following steps before submitting your PR -->
<!--- Feel free to type an `x` in all the boxes below to let us know you have completed the steps: -->
- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)
<!--- You may find this guide helpful: https://mainmatter.com/blog/2021/05/26/keeping-a-clean-git-history/ -->

## Type of change
<!--- Select one of the categories below by typing an `x` in the box -->
- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):
<!--- Additionally, if you are a maintainer or reviewer, please make sure that the appropriate labels are added to this PR -->

ricardoV94 commented:
1. Better to have the same types as return, static shape to constant is introduced during rewrites already 

2 and 3. I would really like to have these, it's what I needed for the batched_dot_to_core rewrites.This isn't a simple case of vectorize because the dims I was  to pack are both on the left and right of other dims

ricardoV94 commented on pytensor/tensor/extra_ops.py:
```diff
@@ -2074,6 +2074,73 @@ def concat_with_broadcast(tensor_list, axis=0):
+    for shape in packed_shapes:
+        size = prod(shape, no_zeros_in_input=True)
+        end = start + size
+        unpacked_tensors.append(
```
Take uses advanced indexing. Actually the best here is probably split. Join and split are the inverses of each other, and will be easier to rewrite away

ricardoV94 commented on pytensor/tensor/extra_ops.py:
```diff
@@ -2074,6 +2074,73 @@ def concat_with_broadcast(tensor_list, axis=0):
+    start = 0
+    unpacked_tensors = []
+    for shape in packed_shapes:
+        size = prod(shape, no_zeros_in_input=True)
```
Why no zeros in input? The shape doesn't show up in gradients if that's what you were worried about

ricardoV94 commented:
I am inclined to making this a core op and not just a helper. It obliviates most uses of reshape and it's much easier to reason about, not having to worry about pesky -1 or whether the reshape shape comes from the original input shapes or not.

That would pretty much address https://github.com/pymc-devs/pytensor/discussions/883

We could use OFG and/or specialize to reshape/split later. It need also not be done in this PR. It's an implementation detail as far as the user is concerned.

jessegrabowski commented on pytensor/tensor/extra_ops.py:
```diff
@@ -2074,6 +2074,73 @@ def concat_with_broadcast(tensor_list, axis=0):
     return join(axis, *bcast_tensor_inputs)
 
 
+def pack(
```
yeah i will of course add docstrings. This was just to get a PR on the board and see your reaction to the 3 issues I raised. I didn't want to document everything before we decided on the final API

jessegrabowski commented:
> I am inclined to making this a core op and not just a helper. It obliviates most uses of reshape and it's much easier to reason about, not having to worry about pesky -1 or whether the reshape shape comes from the original input shapes or not.

I pushed a commit that adds a "feature complete" `Pack` `Op`. It can do everything that `einops.pack` can do, and more. I'd say it's a bit on the overly complex side, but that's on brand for me. The API I cooked up could maybe be simplified. 

Basically, you pack by selecting which axes you don't want to ravel. Axes should be None, int, or tuple[int]. 

If None, it's the same as `packed_tensor = pt.join(0, *[var.ravel() for var in input_vars])`:

```py
x = pt.tensor("x", shape=())
y = pt.tensor("y", shape=(5,))
z = pt.tensor("z", shape=(3, 3))
packed_tensor, packed_shapes = pt.pack(x, y, z, axes=None)
packed_tensor.type.shape # (15,)
```

Once you pass in an integers, your inputs need to have the same shape on the dimensions of concatenation. All dimensions that are in a "hole" of the provided axes is raveled and joined. What's a hole? You can have explicit or implicit holes. An "explicit" hole is a gap in the integers you provide. For example, `[0, -1]` has an explicit hole: all dimensions except the first and last will raveled and joined. 

```py
x = pt.tensor("x", shape=(5, 3))
y = pt.tensor("y", shape=(5, 2, 4, 3))
z = pt.tensor("z", shape=(5, 6, 3))
packed_tensor, packed_shapes = pt.pack(x, y, z, axes=None)
packed_tensor.type.shape # (5, 13, 3)
```

`axes=[0,3]` also looks like an explicit hole -- dimensions 1 and 2 will be raveled and joined. But there's also an *implicit* hole, because there could be dimensions beyond 3. That makes 2 holes, which is an invalid pack. To resolve this, we assume that `3` is the maximum dimension size -- you need to pass at least one tensor with `ndims==4` (since we want to concate on axis=3), and no input can have more. Under those conditions, `axes=[0, 3]` is treated the same as `axes=[0, -1]`:

```py
x = pt.tensor('x', shape=(2, 6))
y = pt.tensor('y', shape=(2, 3, 7, 6))
z = pt.tensor('z', shape=(2, 10, 6))
packed_tensor, packed_shapes = pt.pack(x, y, z, axes=None)
packed_tensor.type.shape # (2, 32, 6)
```

Minimum size is also enforced -- we couldn't have passed in `x = pt.tensor('x', shape=(2,))` for example, because we need at least 2 dimensions to concatenate, because we asked for 2 pack axes.

I could imagine this being more strict -- `axes=[0, 3]` could also imply that **all** inputs are `ndim==4`. The reason I didn't do it this way is because I wanted to match the feel of `einops`. `axis=[0, 3]` feels like it should correspond to something like `i * j`, with *at most* 2 dimensions inside the ellipsis. 

`Pack` could probably also be an `OpFromGraph`, there's nothing special going on in the `perform` method. That would be better because then we get the gradients for free.

ricardoV94 commented on pytensor/tensor/extra_ops.py:
```diff
@@ -2011,6 +2013,345 @@ def concat_with_broadcast(tensor_list, axis=0):
+    pack_op = Pack(
+        inputs=tensors,
+        outputs=[packed_output_tensor, *packed_output_shapes],
+        name="Pack{axes=" + str(axes) + "}",
```
```suggestion
        name="Pack{axes=" + str(axes) + "}",
        inline="True",
```

ricardoV94 commented on pytensor/tensor/extra_ops.py:
```diff
@@ -2011,6 +2013,345 @@ def concat_with_broadcast(tensor_list, axis=0):
+    pack_op = Pack(
+        inputs=tensors,
+        outputs=[packed_output_tensor, *packed_output_shapes],
+        name="Pack{axes=" + str(axes) + "}",
```
Why give name instead of just defining the `__str__` of the `Pack`?

jessegrabowski replied:
Because I looked at the docs for `OpFromGraph` and saw there was a `name` field I could pass

ricardoV94 commented on pytensor/tensor/extra_ops.py:
```diff
@@ -2011,6 +2013,345 @@ def concat_with_broadcast(tensor_list, axis=0):
+
+        if len(set(axes)) != len(axes):
+            raise ValueError("axes must have no duplicates")
+        if axes is not None and len(axes) == 0:
```
No Ops should be supported in general. Makes writing code easier because you don't have to think about edge case. Empty axes are supported in most Ops that allow variable number of axes (like sum)

jessegrabowski commented on pytensor/tensor/extra_ops.py:
```diff
@@ -2011,6 +2013,345 @@ def concat_with_broadcast(tensor_list, axis=0):
+        self.axes = tuple(axes) if isinstance(axes, list) else axes
+        self.op_name = "Pack{axes=" + str(self.axes) + "}"
+
+    def _analyze_axes_list(self) -> tuple[int, int, int, int | None]:
```
technically you asked for it

ricardoV94 replied:
How so?

ricardoV94 commented on pytensor/tensor/extra_ops.py:
```diff
@@ -2267,17 +2267,113 @@ def pack(
+        )  # shape (2, 17)
+
+    `axes` can also be negative, in which case the axes are counted from the end of the tensor shape. For example,
+    if `axes=[-1]`, then the last dimension of each tensor is preserved, and all other dimensions are raveled:
```
The axes specification seems troublesome for vectorization. If axis were those to ravel, then vectorization of this input is very much like axis for other Ops with axes, just shift them to the right by the number of new batch dims.

ricardoV94 replied:
Nvm it's not too bad, we need to add the new batch axis to whatever was defined originally. I'm a bit worried about negative axis at the Op level still

ricardoV94 commented:
Ok 1) is not intuitive, but 2 may still have some merit? It's strictly more powerful and maybe less magical to specify the axes for each input. It also means we can normalize them to be positive and I think simplify the code analysis?

In your use cases this would be terrible UX?

jessegrabowski commented:
The main use-case I have in mind for this is in optimize/pymc where we get parameters in arbitrary shapes, but we want to pack them into a single vector and do a replacement of the original variables with elements of that vector (see the `test_make_replacements_with_pack_unpack` test for what I have in mind).

Having to specify the number of dims ahead of time in that case doesn't work, because we don't know what the user will give us.

ricardoV94 commented:
Why do you need to specify dims before you get to see the inputs?

jessegrabowski commented:
@ricardoV94 I soft rebooted this PR, starting from `join_dims` and `split_dims`. Let me know how those look to you. Next commit will bring back pack/unpack.

The only thing you missed with `unpack_dims` was that the incoming `shapes` can (should?) also be a list, so we need to insert them into the old shape list.

jessegrabowski commented:
I created a new file, `reshape_ops`, and moved `join_dims`, `split_dims`, `pack`, and `unpack` there. I also created new Ops: `JoinDims` and `SplitDims`. I haven't done the rewrites yet, because I'm not confident that these are what you had in mind.

`pack` still doesn't use `join_dims`, because of an important corner case where `pack` needs to create new dimensions. I left a note in the code, but the case is something like:

```py
x = pt.tensor('x', shape=(3,))
y = pt.tensor('y', shape=(3, 5))
z, shapes = pt.pack([x, y], axes=0)
```

To pack this, `x` needs to have a 1 inserted on the right-hand side. The current code does `x.reshape(3, -1)`, which does the insertion. Using `join_dims(x, 0)` will return `(3, )`, which then errors out during `join`.

I'll keep trying, but I'm struggling to see a way to cover all cases.

ricardoV94 commented on pytensor/tensor/shape_ops.py:
```diff
@@ -0,0 +1,461 @@
+
+class JoinDims(Op):
+    def __init__(self, axis: Sequence[int] | int | None = None):
+        self.axis = axis
```
refuse negative axis at the op level (to avoid redundant parametrizations) and raise if non contiguous.

ricardoV94 commented on pytensor/tensor/shape_ops.py:
```diff
@@ -0,0 +1,461 @@
+
+
+class JoinDims(Op):
+    def __init__(self, axis: Sequence[int] | int | None = None):
```
needs props

ricardoV94 commented on pytensor/tensor/shape_ops.py:
```diff
@@ -0,0 +1,461 @@
+            shape=tuple([x if isinstance(x, int) else None for x in output_shapes]),
+            dtype=x.type.dtype,
+        )
+        return Apply(self, [x, as_tensor_variable(shape)], [output])
```
some comments from join dims apply here.

Also assert shape dtype is discrete

ricardoV94 reviewed (commented):
looks good, ops still need some work. right now you're breaking inplace API by not stating the outputs are views

jessegrabowski commented on pytensor/tensor/shape_ops.py:
```diff
@@ -0,0 +1,461 @@
+        (out,) = outputs
+
+        output_shape = [
+            *x.shape[: min(self.axis)],
```
I prefer the axis list. It strikes me as more natural.

ricardoV94 replied:
could also be from axis - to axis. Does it make pack code more simple for instance?

ricardoV94 commented on pytensor/tensor/shape_ops.py:
```diff
@@ -0,0 +1,522 @@
+        self.axis = axis
+
+    def _make_output_shape(self, input_shape, shape):
+        [axis] = normalize_axis_tuple(self.axis, len(input_shape))
```
isn't it already normalized by not allowing negative?

ricardoV94 commented on pytensor/tensor/shape_ops.py:
```diff
@@ -0,0 +1,522 @@
+        def _get_constant_shape(x):
+            try:
+                # get_underling_scalar_constant_value returns a numpy scalar, we need a python int
+                return get_underlying_scalar_constant_value(x).item()
```
this doesn't make sense for perform either, it's a graph extraction operation

ricardoV94 commented on pytensor/tensor/shape_ops.py:
```diff
@@ -0,0 +1,522 @@
+    if all([n_before == 0, n_after == 0, min_axes == 0]):
+        # Special case -- we're raveling everything
+        packed_shapes = [t.shape for t in tensor_list]
+        reshaped_tensors = [t.ravel() for t in tensor_list]
```
why revel instead of join dims? It's still a reshape under the hood. May seem like a simple one, but once we vectorize becomes already ugly

ricardoV94 commented on pytensor/tensor/shape_ops.py:
```diff
@@ -0,0 +1,522 @@
+        # shape[max(axes)+1:]). So this will work if we choose axes=(n_before, n_after_packed - 1). Because of the
+        # rules on the axes input, we will always have n_before <= n_after_packed - 1. A set is used here to cover the
+        # corner case when n_before == n_after_packed - 1 (i.e., when there is only one axis to ravel --> do nothing).
+        join_axes = {n_before, n_after_packed - 1}
```
shouldn't this be a range?

ricardoV94 commented on pytensor/tensor/shape_ops.py:
```diff
@@ -0,0 +1,541 @@
+    view_map = {0: [0]}
+
+    def __init__(self, start_axis: int, n_axes: int):
+        if start_axis < 0:
```
also validate n_axes?

ricardoV94 commented on pytensor/tensor/shape_ops.py:
```diff
@@ -0,0 +1,541 @@
+        return type_cast(TensorVariable, x.squeeze(axis=axis))
+
+    [axis] = normalize_axis_tuple(axis, x.ndim)  # type: ignore[misc]
+    shape = as_tensor_variable(shape, dtype="int64")  # type: ignore[arg-type]
```
```suggestion
    shape = as_tensor_variable(shape, dtype="int64", ndim=1)  # type: ignore[arg-type]
```

ricardoV94 commented on pytensor/tensor/shape_ops.py:
```diff
@@ -0,0 +1,541 @@
+    if positive_axes and not strictly_increasing(positive_axes):
+        raise ValueError("Axes must be strictly increasing in the positive part")
+    if negative_axes and not strictly_increasing(negative_axes):
+        raise ValueError("Axes must be strictly increasing in the negative part")
```
Tiny code reduction?
```suggestion
    def strictly_increasing(s):
        if len(s) < 1:
            return True
        return all(b > a for a, b in pairwise(s))

    if not strictly_increasing(positive_axes):
        raise ValueError("Axes must be strictly increasing in the positive part")
    if not strictly_increasing(negative_axes):
        raise ValueError("Axes must be strictly increasing in the negative part")
```

jessegrabowski replied:
Not sure what you mean with this comment

jessegrabowski replied:
Ok, i simplified that and changed it to an `any` expression too, so that it can eagerly fail the check more quickly if possible

ricardoV94 commented on pytensor/tensor/shape_ops.py:
```diff
@@ -0,0 +1,541 @@
+    if pos_gaps:
+        raise ValueError("Positive axes must be contiguous")
+    if neg_gaps:
+        raise ValueError("Negative axes must be contiguous")
```
You don't  this anywhere? In that case you can change the helper to return a boolean already instead of the "gaps"
```suggestion
    if find_gaps(positive_axes):
        raise ValueError("Positive axes must be contiguous")
    if find_gaps(negative_axes):
        raise ValueError("Negative axes must be contiguous")
```

jessegrabowski commented:
I cleaned up the git history to just 3 commits, each one implementing one bit of this PR. Also addressed the most recent round of feedback. Hopefully we're getting this close to the finish line.

jessegrabowski commented:
Oh, I renamed the files to `reshape.py` and `test_reshape.py`. I thought about moving reshape to `reshape.py`, but that implicated ~70 files so I didn't want to do it in this PR. After we merge this, I'll handle that.

jessegrabowski commented on pytensor/tensor/__init__.py:
```diff
@@ -133,6 +133,11 @@ def _get_vector_length_Constant(op: Op | Variable, var: Constant) -> int:
+
+# isort: off
+# reshape needs to be imported before shape.reshape, otherwise the tensor.reshape imports fail
+from pytensor.tensor.reshape import *
```
There is an `__all__` though

ricardoV94 replied:
missed it
